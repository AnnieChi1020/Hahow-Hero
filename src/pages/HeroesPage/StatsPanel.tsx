import React from 'react';
import styled from 'styled-components';
import Button from '../../components/Button';
import { Stats } from '../../hooks/useProfile';
import { HeroProfile } from '../../redux/heroSlice';
import { H3 } from '../../styles/Fonts';
import StatItem from './StatItem';

interface StatsPanelProps {
    stats: Stats;
    setStats: (stats: Stats) => void;
    handleOnSaveProfile: (profile: HeroProfile) => void;
}

const Container = styled.div`
    display: flex;
    justify-content: space-between;

    @media (max-width: 768px) {
        flex-direction: column;
    }
`;

const RightContainer = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    gap: 20px;

    @media (max-width: 768px) {
        margin-top: 40px;
        h3 {
            text-align: right;
        }
    }
`;

function StatsPanel(props: StatsPanelProps) {
    const { stats, setStats, handleOnSaveProfile } = props;

    const handleIncrement = (type: keyof Stats) => {
        if (stats.rest <= 0) return;
        setStats({
            ...stats,
            [type]: stats[type] + 1,
            rest: stats.rest - 1,
        });
    };

    const handleDecrement = (type: keyof Stats) => {
        if (stats[type] <= 0) return;
        setStats({
            ...stats,
            [type]: stats[type] - 1,
            rest: stats.rest + 1,
        });
    };

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { rest, ...statItems } = stats;

    return (
        <Container>
            <div>
                {Object.keys(statItems).map((stat) => (
                    <StatItem
                        key={stat}
                        label={stat.toUpperCase()}
                        point={stats[stat as keyof HeroProfile]}
                        handleIncrement={() =>
                            handleIncrement(stat as keyof HeroProfile)
                        }
                        handleDecrement={() =>
                            handleDecrement(stat as keyof HeroProfile)
                        }
                        restPoint={stats.rest}
                    />
                ))}
            </div>
            <RightContainer>
                <H3>{`剩餘點數: ${stats.rest}`}</H3>
                <Button onClick={() => handleOnSaveProfile(statItems)}>
                    儲存
                </Button>
            </RightContainer>
        </Container>
    );
}

export default StatsPanel;
