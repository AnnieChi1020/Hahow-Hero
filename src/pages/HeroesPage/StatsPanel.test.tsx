import { render, waitFor } from '@testing-library/react';
import React from 'react';
import StatsPanel from './StatsPanel';

describe('StatsPanel', () => {
    const setStats = jest.fn();
    const handleOnSaveProfile = jest.fn();

    it('should render correctly', () => {
        const stats = { str: 5, int: 2, agi: 8, luk: 3, rest: 0 };

        const { container } = render(
            <StatsPanel
                stats={stats}
                setStats={setStats}
                handleOnSaveProfile={handleOnSaveProfile}
            />
        );

        expect(container).toMatchSnapshot();
    });

    it('should call setStats when click on increment button', () => {
        const stats = { str: 5, int: 2, agi: 8, luk: 3, rest: 0 };

        const { queryAllByRole } = render(
            <StatsPanel
                stats={stats}
                setStats={setStats}
                handleOnSaveProfile={handleOnSaveProfile}
            />
        );

        const buttons = queryAllByRole('button');
        const incrementButton = buttons[1];

        incrementButton.click();

        waitFor(() => {
            expect(setStats).toHaveBeenCalled();
        });
    });

    it('should call setStats when click on decrement button', () => {
        const stats = { str: 5, int: 2, agi: 8, luk: 3, rest: 0 };

        const { queryAllByRole } = render(
            <StatsPanel
                stats={stats}
                setStats={setStats}
                handleOnSaveProfile={handleOnSaveProfile}
            />
        );

        const buttons = queryAllByRole('button');
        const decrementButton = buttons[0];

        decrementButton.click();

        waitFor(() => {
            expect(setStats).toHaveBeenCalled();
        });
    });

    it('should call handleOnSaveProfile when click on save button', () => {
        const stats = { str: 5, int: 2, agi: 8, luk: 3, rest: 0 };

        const { queryAllByRole } = render(
            <StatsPanel
                stats={stats}
                setStats={setStats}
                handleOnSaveProfile={handleOnSaveProfile}
            />
        );

        const buttons = queryAllByRole('button');
        const saveButton = buttons[buttons.length - 1];

        saveButton.click();

        waitFor(() => {
            expect(handleOnSaveProfile).toHaveBeenCalled();
        });
    });

    it('should not call handleOnSaveProfile when click on save button and restPoint is not zero', () => {
        const stats = { str: 5, int: 2, agi: 8, luk: 3, rest: 1 };

        const { queryAllByRole } = render(
            <StatsPanel
                stats={stats}
                setStats={setStats}
                handleOnSaveProfile={handleOnSaveProfile}
            />
        );

        const buttons = queryAllByRole('button');
        const saveButton = buttons[buttons.length - 1];

        saveButton.click();

        waitFor(() => {
            expect(handleOnSaveProfile).not.toHaveBeenCalled();
        });
    });
});
