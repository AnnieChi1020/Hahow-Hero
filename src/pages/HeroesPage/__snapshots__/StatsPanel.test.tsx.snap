// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsPanel should render correctly 1`] = `
<div>
  <div
    class="sc-jsMahE kAzAbn"
  >
    <div
      class="sc-kFuwaP"
    >
      <div
        class="sc-gLDzan IUsCp"
      >
        <h3
          class="sc-hLseeU sc-iAEyYk jaQnIa gpwMbS"
        >
          STR
        </h3>
        <button
          class="sc-beqWaB gRbeAN"
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            -
          </h3>
        </button>
        <h3
          class="sc-hLseeU sc-fsQiph jaQnIa gYRmdc"
        >
          5
        </h3>
        <button
          class="sc-beqWaB cmFagU"
          disabled=""
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            +
          </h3>
        </button>
      </div>
      <div
        class="sc-gLDzan IUsCp"
      >
        <h3
          class="sc-hLseeU sc-iAEyYk jaQnIa gpwMbS"
        >
          INT
        </h3>
        <button
          class="sc-beqWaB gRbeAN"
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            -
          </h3>
        </button>
        <h3
          class="sc-hLseeU sc-fsQiph jaQnIa gYRmdc"
        >
          2
        </h3>
        <button
          class="sc-beqWaB cmFagU"
          disabled=""
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            +
          </h3>
        </button>
      </div>
      <div
        class="sc-gLDzan IUsCp"
      >
        <h3
          class="sc-hLseeU sc-iAEyYk jaQnIa gpwMbS"
        >
          AGI
        </h3>
        <button
          class="sc-beqWaB gRbeAN"
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            -
          </h3>
        </button>
        <h3
          class="sc-hLseeU sc-fsQiph jaQnIa gYRmdc"
        >
          8
        </h3>
        <button
          class="sc-beqWaB cmFagU"
          disabled=""
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            +
          </h3>
        </button>
      </div>
      <div
        class="sc-gLDzan IUsCp"
      >
        <h3
          class="sc-hLseeU sc-iAEyYk jaQnIa gpwMbS"
        >
          LUK
        </h3>
        <button
          class="sc-beqWaB gRbeAN"
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            -
          </h3>
        </button>
        <h3
          class="sc-hLseeU sc-fsQiph jaQnIa gYRmdc"
        >
          3
        </h3>
        <button
          class="sc-beqWaB cmFagU"
          disabled=""
        >
          <h3
            class="sc-hLseeU sc-qRumB jaQnIa dNNzgk"
          >
            +
          </h3>
        </button>
      </div>
    </div>
    <div
      class="sc-irTswW cXMkzs"
    >
      <h2
        class="sc-dmqHEX iDIKHP"
      >
        剩餘點數: 0
      </h2>
      <button
        class="sc-beqWaB gRbeAN"
      >
        儲存
      </button>
    </div>
  </div>
</div>
`;
